### ISSUE TITLE
Error resolving module version for dependencies

### ISSUE THREAD

**Issue by zetashift** [projen/projen#2641]
When running `npx projen` in my mono repo I'm getting the following error:
```
foo main• ↑4 ❱ npx projen                                         devenv-shell-env
👾 default | ts-node --project tsconfig.dev.json .projenrc.ts
👾 unable to resolve version for api-typescript-runtime from installed modules
👾 unable to resolve version for api-typescript-infra from installed modules
👾 unable to resolve version for api-typescript-runtime from installed modules
```

- Using projen v0.71.31

### Steps to reproduce

Create a new nx-monorepo

```
mkdir repo && npx projen new --from @aws-prototyping-sdk/nx-monorepo
```

Change the `project` variable in `.projenrc.ts` to this:

```typescript
const project = new NxMonorepoProject({
  defaultReleaseBranch: "main",
  devDeps: ["@aws-prototyping-sdk/nx-monorepo", "@aws-prototyping-sdk/type-safe-api"],
  name: "foo",
  projenrcTs: true,

  deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,  /* The "name" in package.json. */
});
```

Then run `npx projen`.

Then change the .projenrc.ts to:

```typescript
import { NxMonorepoProject } from "@aws-prototyping-sdk/nx-monorepo";
import { AwsCdkTypeScriptApp } from "projen/lib/awscdk";
import {
  DocumentationFormat,
  Language,
  ModelLanguage,
  TypeSafeApiProject,
} from "@aws-prototyping-sdk/type-safe-api";

const project = new NxMonorepoProject({
  defaultReleaseBranch: "main",
  devDeps: ["@aws-prototyping-sdk/nx-monorepo", "@aws-prototyping-sdk/type-safe-api"],
  name: "foo",
  projenrcTs: true,

  deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,  /* The "name" in package.json. */
});


/** Contains all the smithy model files and codegenerates documentation, clients and API types.
 */
const api = new TypeSafeApiProject({
  name: "api",
  parent: project,
  outdir: "packages/api",
  runtime: {
    languages: [Language.TYPESCRIPT],
  },
  infrastructure: {
    language: Language.TYPESCRIPT,
  },
  documentation: {
    formats: [DocumentationFormat.HTML_REDOC],
  },
  model: {
    language: ModelLanguage.SMITHY,
    options: {
      smithy: {
        serviceName: {
          namespace: "foo.api",
          serviceName: "FooApi",
        },
      },
    },
  },
});

/** Our CDK based project that will deploy our API on AWS Lambda.
 * Also contains all the code for the Lambda handlers.
 */
new AwsCdkTypeScriptApp({
  name: "infra",
  parent: project,
  outdir: "packages/infra",
  defaultReleaseBranch: "main",
  cdkVersion: "2.76.0",
  deps: [
    api.runtime.typescript!.package.packageName,
    api.infrastructure.typescript!.package.packageName,
    "node-fetch",
  ],
  devDeps: [
  ],
  eslint: false,
  gitignore: ["edgeql-js"],
});

project.synth();
```

And run `npx projen` again. It should fail to resolve the dependencies of the `AwsCdkTypeScriptApp`

**Comment by mrgrain** [comment]
Likely related to https://github.com/projen/projen/pull/2634

**Comment by mrgrain** [comment]
So more details. This happens when workspaces are used, like yarn and @aws-prototyping-sdk/nx-monorepo do. The reason this happens is because with workspaces dependencies are not installed in the subproject's outdir, but hoisted to the monorepo root.

Now for the fun bit. As far as I can tell this does not correctly work in @aws-prototyping-sdk/nx-monorepo. In the reproduction you've provided, using `projen@0.71.28`  I can see how the `*` are not actually removed. E.g. packages/infra/package.json still contains "jest": "*". Can you confirm this is the same in your real-world project?

From that perspective having the warning makes sense. It was a bug that the warning did not appear before. It wouldn't really be up to you to fix this, but something nx-monorepo needs to do.

Here it get's weird. After the change, projen correctly warns about not being able to resolve the versions - but than goes and resolves them anyway!

So in summary, the situation is:
projen `<= 0.71.29`: No warning, versions are not resolved
projen `>= 0.71.30`: Warning, but versions are actually resolved.

**Comment by zetashift** [comment]
From my real-world project where the root package json contains: `"projen": "^0.71.21", ` my `infra` subprojects package.json contains:

```json
  "devDependencies": {
    "@types/jest": "^29.5.1",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "aws-cdk": "^2.76.0",
    "esbuild": "^0.17.17",
    "eslint": "^8",
    "eslint-import-resolver-node": "^0.3.7",
    "eslint-import-resolver-typescript": "^3.5.5",
    "eslint-plugin-import": "^2.27.5",
    "jest": "^29.5.0",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "projen": "^0.71.20",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5"
  },
  ```

But here is a `frontend` subproject that does contain `*`:

```
  "devDependencies": {
    "@sveltejs/adapter-auto": "*",
    "@sveltejs/kit": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "eslint": "^8",
    "eslint-import-resolver-node": "^0.3.7",
    "eslint-import-resolver-typescript": "^3.5.5",
    "eslint-plugin-import": "^2.27.5",
    "npm-check-updates": "^16",
    "projen": "^0.71.20",
    "svelte": "*",
    "svelte-check": "*",
    "ts-node": "^10.9.1",
    "tslib": "^2.5.0",
    "typescript": "^4.9.5",
    "vite": "^4.3.1"
  },
  ```

**Comment by mrgrain** [comment]
Okay, looking at this again, using the repro. With the latest `projen` version, any subproject dependencies with `*` are removed from `package.json`. With `projen@0.71.28` they stay in there untouched with `*` and all.

Now I think the bug is that unresolved versions are removed while they shouldn't be. How the `infra` subproject ended up with `"jest": "^29.5.0"` in your project I can't really tell. It makes sense though that once `package.json` has a version, it will stay on it.

**Comment by zetashift** [comment]
>  It makes sense though that once package.json has a version, it will stay on it. 

Yup I think this is exactly what happened!
The svelte thing got added last week, and there is a `*` there.


> It wouldn't really be up to you to fix this, but something nx-monorepo needs to do.

Does this mean I should open up an issue on the PDK repo/PR for the PDK? Or should PDK upgrade to a more recent projen version (of which there is a PR open)? Not sure what actionable thing I can do.

**Comment by mrgrain** [comment]
> Yup I think this is exactly what happened!
The svelte thing got added last week, and there is a * there.

I reckon `projen upgrade` might also replace `*`  with a proper version.
That could be how the other packages ended up with it.

> Does this mean I should open up an issue on the PDK repo/PR for the PDK? Or should PDK upgrade to a more recent projen version (of which there is a PR open)? Not sure what actionable thing I can do.

I do think we need an issue with PDK eventually. It's clearly not correct that these versions can't be resolved.

However currently it looks to me like in `projen` it was originally intended to remove dependencies that cannot be resolved.
But due to a bug that just never happened.  And that bug got (unintentionally) fixed last week.

What I'm wondering now is if removing is a good thing, or if warning and keeping them is actually the better approach in the first place.
This would effectively codify the previous behavior as correct.

**Comment by mrgrain** [comment]
It pretty much comes down to this line:
https://github.com/aws/aws-prototyping-sdk/blob/cf57586f2567a571c734606d6f003f6a5051bfca/packages/nx-monorepo/src/nx-monorepo.ts#L766

projen core now (correctly) expects that subproject dependencies are installed before `resolveDepsAndWritePackageJson` is called. Overwriting `installDependencies` on the subproject has been the go to way for monorepos to do this. But of course there's a reason that method is private.

I think what this highlights is the missing support for workspace-based monorepos in `NodeProject`.
I wonder what the API for this should look like.

**Comment by mrgrain** [comment]
tl;dr @zetashift please go ahead an open an issue with PDK as well. Tag me on it and link them to this issue.
I'll have a think about an API and required changes as well.

**Comment by JeremyJonas** [comment]
@agdimech Can you look into this

**Comment by BradenM** [comment]
Thanks to all who are looking into this.
I (ironically) ran into this on my own nx-monorepo based projen project after updating...

While looking into this myself, I found another issue relating to module entry points that could cause some additional confusion while resolving this. See #2644 (and proposed fix: #2645)

**Comment by mrgrain** [comment]
@zetashift @JeremyJonas @BradenM 
Can you confirm if the latetw PDK / `nx-monorepo` release resolves this issue?

**Comment by zetashift** [comment]
For me, the warnings/errors about module resolution disappeared!

**Comment by BradenM** [comment]
@mrgrain Sorry for the late response, missed the notification.
I can confirm that the latest nx-monorepo resolves this issue for me as well.

Just to clarify though, it does not resolve #2644 (which is a separate, actually `projen` issue, but gives the same "👾 [components] unable to resolve version for rollup from installed modules" error still).

**Comment by mrgrain** [comment]
Thanks @BradenM I'll look at the other issue as well soon.
